compiler: gate `extern "{abi}"` in ast_lowering

workingjubilee opened the pull request:
I don't believe low-level crates like `rustc_abi` should have to know or care about higher-level concerns like whether the ABI string is stable for users. These implementation details can be made less open to public inspection. This way the code that governs stability is near the code that enforces stability, and compiled together.

It also abstracts away certain error messages instead of constantly repeating them.

A few error messages are simply deleted outright, instead of made uniform, because they are either too dated to be useful or redundant with other diagnostic improvements we could make. These can be pursued in followups: my first concern was making sure there wasn't unnecessary diagnostics-related code in `rustc_abi`, which is not well-positioned to understand what kind of errors are going to be generated based on how it is used.

r? @ghost

workingjubilee committed:
compiler: gate `extern "{abi}"` in ast_lowering

By moving this stability check into AST lowering, we effectively make
it impossible to accidentally miss, as it must happen to generate HIR.
Also, we put the ABI-stability code next to code that actually uses it!
This allows code that wants to reason about backend ABI implementations
to stop worrying about high-level concerns like syntax stability,
while still leaving it as the authority on what ABIs actually exist.

It also makes it easy to refactor things to have more consistent errors.
For now, we only apply this to generalize the existing messages a bit.

workingjubilee committed:
tests: error strings for ABI stability now match

workingjubilee committed:
compiler: start using rustc_ast_lowering in rustc_passes

fbstj commented on compiler/rustc_ast_lowering/src/stability.rs:
```diff
@@ -0,0 +1,151 @@
+
+#[allow(rustc::untranslatable_diagnostic)]
+pub(crate) fn gate_unstable_abi(sess: &Session, features: &Features, span: Span, abi: ExternAbi) {
+    let result = extern_abi_stability(abi).or_else(|unstable @ UnstableAbi { feature, .. }| {
+        if features.enabled(feature) || span.allows_unstable(feature) {
+            Ok(())
+        } else {
+            Err(unstable)
+        }
+    });
+    match result {
```
```suggestion
    match extern_abi_enabled(abi) {
```
?

workingjubilee replied:
...that's correct! 😵‍💫

workingjubilee replied:
thanks!

workingjubilee committed:
compiler: remove rustc_target::spec::abi reexports

workingjubilee commented:
r? @compiler-errors or reroll

workingjubilee commented:
I split out the first four commits again into #136706 but that's as far as things can be cut.

workingjubilee committed:
compiler: remove `abi`-specific `extern "{abi}"` suggestions

These are either residue of a long-term migration away from something,
or are simply trying too hard to be specifically useful:
nearest-match suggestions for ABI strings should handle this.

bors commented:
:umbrella: The latest upstream changes (presumably #136762) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented on tests/ui/intrinsics/incorrect-transmute.stderr:
```diff
@@ -1,4 +1,4 @@
-error[E0658]: intrinsics are subject to change
+error[E0658]: rust-intrinsic ABI is an implementation detail and perma-unstable
```
nit, what about:

> the "rust-intrinsic" ABI is [...]

compiler-errors replied:
and similarly for above

workingjubilee replied:
Sure! should it be `extern "abi"` so it looks like what people actually write in syntax?

workingjubilee replied:
pushed a diff for consideration~

compiler-errors commented on tests/ui/abi/removed-wasm-abi.rs:
```diff
@@ -1,4 +0,0 @@
-extern "wasm" fn test() {}
```
why remove this test?

workingjubilee replied:
It's a suggestion that was for migrating off a nightly feature, starting from last July.

The only users of the feature that I'm aware of are people implementing next-generation wasm toolchain components. So, the people working with us on the migration of wasm's `extern "C"` to match clang's. They... have known about this for some time, so it seems unlikely they will still benefit from a continued reminder into the future.

So, I removed the suggestion, as we don't usually keep around migration stuff for nightly things for more than a couple of releases, and it's been some.

compiler-errors reviewed (approved):
LGTM, i have a few questions but r=me regardless

workingjubilee commented on tests/ui/abi/riscv-discoverability-guidance.riscv32.stderr:
```diff
@@ -8,10 +8,9 @@ LL | extern "riscv-interrupt" fn isr() {}
    |        help: did you mean: `"riscv-interrupt-m"`
    |
    = note: invoke `rustc --print=calling-conventions` for a full list of supported calling conventions
-   = note: please use one of riscv-interrupt-m or riscv-interrupt-s for machine- or supervisor-level interrupts, respectively
```
It's functionally redundant with this one:
>    |        help: did you mean: `"riscv-interrupt-m"`

If there are two suggestions of equal distance and we want to suggest "try either of these", then we should just do that in a generalized way instead.

workingjubilee commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit cd9d39e3604680db58acbe409590d495e883d053 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors cd9d39e3604680db58acbe409590d495e883d053 -->
<!-- homu: {"type":"Approved","sha":"cd9d39e3604680db58acbe409590d495e883d053","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented on compiler/rustc_ast_lowering/src/stability.rs:
```diff
@@ -0,0 +1,147 @@
+    }
+}
+
+pub fn extern_abi_stability(abi: ExternAbi) -> Result<(), UnstableAbi> {
```
We have that issue for all stability concerns, because we have a really bad organization of how stability checks work.
